### PR TITLE
Clean up `@swc/plugin-relay`

### PR DESCRIPTION
**Pull request by XiNiHa**
- Remove options `src` and `pagesDir` since they're not used or NextJS-specific.
- Fix the logs a bit
- Provide TS typedefs for defining the config

### CODE REVIEW

**Comment by kwonoj** [comment]
I think we can remove `pagesDir` altogether? (https://github.com/swc-project/plugins/issues/35#issuecomment-1131173865=) it was mostly my fault while hurry to copy-paste from next.js.

File: packages/relay/package.json [hunk 1]
```diff
@@ -3,8 +3,15 @@
   "version": "0.1.2",
   "description": "SWC plugin for relay",
   "main": "swc_plugin_relay.wasm",
+  "types": "./dist/index.d.ts",
+  "exports": {
```
Comment (kwonoj) [inline packages/relay/package.json:7]: package.json entrypoint should resolve to actual wasm binary, `@swc/core` uses this to read plugin to execute. pointing this to .js will fail to load plugin.

File: packages/relay/package.json [hunk 2]
```diff
@@ -3,8 +3,15 @@
...
+    "import": "./dist/index.mjs",
+    "require": "./dist/index.js",
+    "default": "./dist/index.js"
+  },
   "scripts": {
-    "prepack": "cp ../../target/wasm32-wasi/release/swc_plugin_relay.wasm ."
+    "build": "tsup index.ts --format cjs,esm --dts",
```
Comment (kwonoj) [inline packages/relay/package.json:13]: What's usecase for tsup in here? Specifically, what's major problem cannot be resolved with plain tsc for this relatively small package?

File: packages/relay/package.json (hunk 1 above)
Comment (XiNiHa) [reply packages/relay/package.json:7]: If `@swc/core` can resolve conditional exports, how about defining a custom exports flag for swc plugins? So plugin authors will be able to publish it with something like `"exports": { "swc-plugin": "swc_plugin_relay.wasm" }`.

File: packages/relay/package.json (hunk 2 above)
Comment (XiNiHa) [reply packages/relay/package.json:13]: It was for dual packaging, but I can change it to only use tsc and then make the package CJS only if you want.

File: packages/relay/package.json (hunk 1 above)
Comment (kwonoj) [reply packages/relay/package.json:7]: I'm not sure to have custom, non-standard field in exports definition. While it's not probably spec'ed, I'd rather to stay with common convention. Secondly, what's actual usecase if we allow to specify entrypoint to plugin in js? there's no way to import / execut plugin package with cjs in normal node.js environment.

File: packages/relay/package.json (hunk 2 above)
Comment (kwonoj) [reply packages/relay/package.json:13]: Before changing this, I think we need to resolve https://github.com/swc-project/plugins/pull/38#discussion_r880045731 question first: who can actually import in cjs / esm in normal node.js environment and how?

File: packages/relay/package.json (hunk 1 above)
Comment (XiNiHa) [reply packages/relay/package.json:7]: > I'm not sure to have custom, non-standard field in exports definition.

Node.js spec doesn't restrict this, and lots of packages are defining edgy custom exports like `deno`, `workers`, etc.

> Secondly, what's actual usecase if we allow to specify entrypoint to plugin in js?

It's for the helper function, not the actual plugin. so users can do this(if they can pass the options dynamically, when using [`unplugin-swc`](https://github.com/egoist/unplugin-swc) for example):

```ts
import relay from '@swc/plugin-relay'

export default {
  jsc: {
    experimental: {
      plugins: [
        relay({
          rootDir: '.', // `resolve()`d inside `relay()`
          language: 'typescript', // typechecked
        })
      ]
    }
  }
}
```

File: packages/relay/package.json (hunk 2 above)
Comment (XiNiHa) [reply packages/relay/package.json:13]: I left a comment there :)

File: packages/relay/package.json (hunk 1 above)
Comment (kwonoj) [reply packages/relay/package.json:7]: For this case, I guess current PR may work.

However, for now I'd like to object this approach, as this is only accessible when using `@swc/core` via node.js runtime interfaces. Cli, or using static .swcrc, or `@swc/wasm` will not able to use this plugin with those functions.

One of the biggest blocker for stablizing plugin is setting right interfaces provide way to load across all of the runtimes swc supports. For the same reason .swcrc does not support any dynamic configuration yet.

Comment (XiNiHa) [reply packages/relay/package.json:7]: Yeah I understand that, and that's why I thought this can't be a standard solution for problems like alternative package manager resolutions.

Comment (XiNiHa) [reply packages/relay/package.json:7]: TBH I think not adapting dynamic features only because of the CLI doesn't sound like a good idea, and `.swcrc` itself should support dynamic configs.

Comment (kwonoj) [reply packages/relay/package.json:7]: I didn't say cli is the only runtime prevents js though.

Comment (XiNiHa) [reply packages/relay/package.json:7]: What are the other runtimes blocking dynamic JS config? If `.swcrc` can be dynamic and the resulting/evaluated config can be serialized, is there any reason why it couldn't be used in other runtimes like `@swc/wasm`?

Comment (kwonoj) [reply packages/relay/package.json:7]: I believe this PR is not suitable for the discussion as it's large topic we discuss internally, anyway for the short summary - first, when user typically want `js` based dynamic they usually expect whole ecosystem - mostly based on node.js. This is not entirely true for the swc, as we expect to run this over v8-based runtime other than node.js, including deno but also any edge workers based on v8. In that case, the scope of `js` becomes really narrow already and there can be lot of confusion / conflict. But moreover, there'll be other runtimes we won't able to use js even. Cli is one, but more importantaly `@swc/wasm` on the direct wasm runtime outside of v8. In this case none of js can be executed since wasm runtime is not tied to certain js runtime.

In short, one of the major goal of the SWC is agnostic to certain runtime as much and even if we support dynamic way to configure (I believe we'll) it's likely not be js, or specifically node.js. Of course major usecase of swc for now is running it via node.js with napi, but that doesn't mean that covers all of the usecase.

Comment (XiNiHa) [reply packages/relay/package.json:7]: Gotcha, do you want me to remove all the helpers and only leave a `.d.ts` file then?

Comment (kwonoj) [reply packages/relay/package.json:7]: I think so, For now it'd be great if you can align with existing plugins by using main field to point .wasm binary. I bet we'll have breaking changes for this once we're out of experimental phase, but at least we can consistently apply changes at once.

**Review by kwonoj — APPROVED** [review]

**Review by kdy1 — APPROVED** [review]
